bindings/java/src/org/sleuthkit/datamodel/OsAccountManager.java

```diff
@@ -0,0 +1,653 @@
+	 * @param dataSourceObjId Object id of the data source where the instance is found.
+	 * @param instanceType Instance type.
+	 * @param connection   Database connection to use.
+	 *
+	 * @throws TskCoreException
+	 */
+	void addOsAccountInstance(OsAccount osAccount, Host host, long dataSourceObjId, OsAccount.OsAccountInstanceType instanceType, CaseDbConnection connection) throws TskCoreException {
+
+		db.acquireSingleUserCaseWriteLock();
+		try {
+			String userInsertSQL = "INSERT INTO tsk_os_account_instances(os_account_id, data_source_obj_id, host_id, instance_type)"
+					+ " VALUES (?, ?, ?, ?)"; // NON-NLS
+
+			PreparedStatement preparedStatement = connection.getPreparedStatement(userInsertSQL, Statement.RETURN_GENERATED_KEYS);
+			preparedStatement.clearParameters();
+
+			preparedStatement.setLong(1, osAccount.getId());
+			preparedStatement.setLong(2, dataSourceObjId);
+			preparedStatement.setLong(3, host.getId());
+			preparedStatement.setInt(4, instanceType.getId());
+			
+			connection.executeUpdate(preparedStatement);
+			
+		} catch (SQLException ex) {
+			LOGGER.log(Level.SEVERE, null, ex);
+			throw new TskCoreException(String.format("Error adding os account instance for account = %s, host name = %s, data source object id = %d", osAccount.getSignature(), host.getName(), dataSourceObjId ), ex);
+		} finally {
+			db.releaseSingleUserCaseWriteLock();
+		}
+	}
+	
+	/**
+	 * Get all accounts that had an instance on the specified host.
+	 * 
+	 * @param host Host for which to look accounts for.
+	 * 
+	 * @return Set of OsAccounts, may be empty.
+	 */
+	Set<OsAccount>  getAcccounts(Host host) throws TskCoreException {
+	
+		String queryString = "SELECT * FROM tsk_os_accounts as accounts "
+				+ " JOIN tsk_os_account_instances as instances "
+				+ " ON instances.os_account_id = accounts.obj_id "
+				+ " WHERE instances.host_id = " + host.getId();
+
+		try (CaseDbConnection connection = this.db.getConnection();
+				Statement s = connection.createStatement();
+				ResultSet rs = connection.executeQuery(s, queryString)) {
+
+			Set<OsAccount> accounts = new HashSet<>();
+			while (rs.next()) {
+				OsAccountRealm realm = null;
+				long realmId = rs.getLong("realm_id");
+				if (!rs.wasNull()) {
+					realm = db.getOsAccountRealmManager().getRealm(realmId, connection);
+				}
+
+				accounts.add(osAccountFromResultSet(rs, realm));
+			} 
+			return accounts;
+		} catch (SQLException ex) {
+			throw new TskCoreException(String.format("Error getting OS accounts for host id = %d", host.getId()), ex);
+		}
+	}
+	
+	
+	/**
+	 * Gets or creates an OS account with the given SID.
+	 * 
+	 * @param sid Account SID.
+	 * @param host Host for the realm.
+	 * @param transaction Transaction to use.
+	 * 
+	 * @return OsAccount for the user.
+	 * @throws TskCoreException 
+	 */
+	OsAccount getOrCreateOsAccountByWindowsSID(String sid, Host host, CaseDbTransaction transaction) throws TskCoreException {
```

Comment:
@bcarrier 
This method has a check then act race condition for multi-user cases or in multithreaded scenarios.

What if the user does not have a Windows SID?

Reply:
Check then act race conditions have been addressed by separating methods. 

There is another API to create account by domain/loginName.